external/dhcpd/dhcpd_lwip.c

```diff
@@ -103,29 +102,113 @@ typedef struct dhcp_join_data dhcp_join_data_s;
 #define DHCPD_MQ_LEN 11
 #define DHCPD_MQ_MAX_LEN 20
 
-void *_dhcpd_join_handler(void *arg)
+static int _send_mqmsg(mqd_t mq, char *msg, size_t msglen, int prio)
 {
-	dhcp_join_data_s *data = (dhcp_join_data_s *)arg;
-	int ret = ERROR;
-	int sockfd = 0;
-	struct req_lwip_data req;
+	nvdbg("dhcps send dhcps msg size %d\n", msglen);
+	while (1) {
+		int mq_ret = mq_send(mq, msg, msglen, prio);
+		if (mq_ret == -1) {
+			if (errno == EAGAIN || errno == EINTR) {
+				continue;
+			}
+			ndbg("dhcps _send_mqmsg error %d\n", errno);
+			return -1;
+		}
+		nvdbg("dhcps sent res %d sent %d\n", mq_ret, msglen);
+		break;
+	}
+	return 0;
+}
 
+static int _recv_mqmsg(mqd_t mq, char *msg, size_t msglen, int *prio)
+{
+	nvdbg("dhcps receive dhcps msg size %d\n", msglen);
+	while (1) {
+		int mq_ret = mq_receive(mq, msg, msglen, prio);
+		if (mq_ret == -1) {
+			if (errno == EAGAIN && errno == EINTR) {
+				continue;
+			}
+			ndbg("dhcps _recv_mqmsg error %d\n", errno);
+			return -1;
+		}
+		if (mq_ret == msglen) {
+			break;
+		} else {
+			ndbg("dhcps critical error %d %d\n", mq_ret, msglen);
+		}
+	}
+	return 0;
+}
+
+static void _empty_mq(mqd_t mq)
+{
+	nvdbg("dhcps empty dhcps message queue\n");
+	char msg[DHCPD_MQ_LEN];
+	struct timespec ts;
+	int status = clock_gettime(CLOCK_REALTIME, &ts);
+	if (status != 0) {
+		ndbg("dhcps ERROR clock_gettime failed\n");
+	}
+	ts.tv_nsec += 5000000; // 5 milli seconds
+	while (1) {
+		struct mq_attr mq_stat;
+		mq_getattr(mq, &mq_stat);
+		if (mq_stat.mq_curmsgs == 0) {
+			nvdbg("dhcps dhcps msg queue is empty\n");
+			return;
+		}
+		int nbytes = mq_timedreceive(mq, msg, DHCPD_MQ_LEN, 0, &ts);
```

Comment:
In every loop, it consumes the time. So most of the cases are okay, but sometimes it cannot wait 5ms accurately.

Reply:
If there're many messages then mq_timedreceive will return message immediately. So I think 5ms is ok. and 5ms doesn't affect to system performance. If you have any idea please let me know.

Reply:
I also think 5ms is too short to read all pending messages.
Otherwise, why don't you move the timer setting logic into the loop.

Reply:
I also think we'd better set the interval in the loop.

Reply:
it only returns when ETIMEOUT is returned. if mq_timedreceive() returns message than it tries to read mq message again.

Reply:
As discussed on-line, please move time setting into the loop.